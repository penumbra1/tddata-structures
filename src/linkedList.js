export function createNode(value) {
  return {
    value,
    next: null
  };
}

// Singly linked list
export function createLinkedList() {
  return {
    head: null,
    tail: null,
    length: 0,
    get isEmpty() {
      return this.length === 0;
    },
    get(index) {
      if (index < 0 || index > this.length - 1) return null;

      if (index === 0) return this.head;
      if (index === this.length - 1) return this.tail;

      let current = this.head;

      let i = 0;
      while (i !== index) {
        current = current.next;
        i++;
      }

      return current;
    },
    delete(index) {
      if (index < 0 || index > this.length - 1) return null;

      let currentIndex = 0;
      let [previous, current, following] = [null, this.head, this.head.next];

      while (currentIndex < index) {
        [previous, current, following] = [current, following, following.next];
        currentIndex++;
      }

      if (previous === null) {
        // Current head will be removed
        this.head = following;
      } else {
        previous.next = following;
      }

      if (following === null) {
        // Current tail will be removed
        this.tail = previous;
      }

      this.length--;

      return current;
    },
    push(value) {
      const node = createNode(value);

      if (this.length === 0) {
        this.head = node;
      } else {
        this.tail.next = node;
      }

      this.tail = node;
      this.length++;
    },
    pop() {
      if (this.length === 0) {
        return null;
      }

      const oldTail = this.tail;
      // newTail will refer to the node previous to oldTail
      let newTail = this.head;

      if (this.length === 1) {
        this.head = null;
        newTail = null;
      } else {
        while (newTail.next !== this.tail) {
          newTail = newTail.next;
        }

        newTail.next = null;
      }

      this.tail = newTail;
      this.length--;

      return oldTail;
    },
    shift(value) {
      const node = createNode(value);

      node.next = this.head;
      this.head = node;
      if (this.length === 0) {
        this.tail = node;
      }
      this.length++;
    },
    unshift() {
      if (this.length === 0) {
        return null;
      }

      if (this.length === 1) {
        this.tail = null;
      }

      const oldHead = this.head;
      this.head = this.head.next;
      this.length--;

      return oldHead;
    },
    insertAt(index, value) {
      const node = createNode(value);

      if (index < 0 || index > this.length) return null;

      let currentIndex = 0;
      let [insertAfter, insertBefore] = [null, this.head];

      if (index === this.length) {
        // No need to traverse the list - fast forward to the end:
        currentIndex = this.length;
        [insertAfter, insertBefore] = [this.tail, null];
        this.tail = node;
      }

      while (currentIndex < index) {
        [insertAfter, insertBefore] = [insertBefore, insertBefore.next];
        currentIndex++;
      }

      if (insertAfter === null) {
        this.head = node;
      } else {
        insertAfter.next = node;
      }

      node.next = insertBefore;

      this.length++;

      return node;
    },
    find(value) {
      let current = this.head;
      let index = 0;

      while (current && current.value !== value) {
        current = current.next;
        index++;
      }

      if (current === null) return null;

      return index;
    },
    print() {
      let current = this.head;
      const values = [];

      while (current) {
        values.push(current.value);
        current = current.next;
      }

      return values.join(" => ");
    }
  };
}
