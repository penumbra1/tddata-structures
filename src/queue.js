export function createQueue() {
  const queue = [];

  return {
    enqueue(item) {
      queue.unshift(item);
    },
    dequeue() {
      return queue.pop();
    },
    peek() {
      return queue[queue.length - 1];
    },
    get length() {
      return queue.length;
    },
    get isEmpty() {
      return queue.length === 0;
    }
  };
}

export function createPriorityQueue() {
  const lowPriority = createQueue();
  const highPriority = createQueue();
  return {
    enqueue(item, isHighPriority = false) {
      if (isHighPriority) highPriority.enqueue(item);
      else lowPriority.enqueue(item);
    },
    dequeue() {
      if (!highPriority.isEmpty) return highPriority.dequeue();
      return lowPriority.dequeue();
    },
    peek() {
      if (!highPriority.isEmpty) return highPriority.peek();
      return lowPriority.peek();
    },
    get length() {
      return highPriority.length + lowPriority.length;
    },
    get isEmpty() {
      return highPriority.isEmpty && lowPriority.isEmpty;
    }
  };
}
