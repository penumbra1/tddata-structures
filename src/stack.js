export default function createStack(max = 5) {
  const stack = [];

  return {
    max,
    push(item) {
      if (stack.length < max) stack.push(item);
      else throw new Error("Stack overflow!");
    },
    pop() {
      return stack.pop();
    },
    peek() {
      return stack[stack.length - 1];
    },
    get isEmpty() {
      return stack.length === 0;
    },
    get length() {
      return stack.length;
    }
  };
}
