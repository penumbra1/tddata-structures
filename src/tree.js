export function createNode(key) {
  const children = [];

  return {
    key,
    children,
    addChild(childKey) {
      const child = createNode(childKey);
      children.push(child);
      return child;
    }
  };
}

export function createTree(rootKey) {
  if (rootKey == null) throw new Error("Creating a tree requires a root key");

  const root = createNode(rootKey);

  return {
    root,
    traverse(node, action, depth) {
      action(node, depth);

      if (node.children.length) {
        node.children.forEach(child => {
          this.traverse(child, action, depth + 1);
        });
      }
    },
    print() {
      let result = "";
      const addLine = (node, depth) => {
        result += depth
          ? `${"|  ".repeat(depth)}|--${node.key}\n`
          : `|--${node.key}\n`;
      };

      this.traverse(this.root, addLine, 0);

      return result;
    }
  };
}
