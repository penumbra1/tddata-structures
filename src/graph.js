import { createQueue } from "./queue";

export function createNode(key) {
  const neighbors = [];

  return {
    key,
    neighbors,
    addNeighbor(node) {
      neighbors.push(node);
    },
    print() {
      const neighborList = neighbors.map(n => n.key).join(" ");
      const line = `${this.key}${neighborList && ` => ${neighborList}`}`;

      return line;
    }
  };
}

export function createGraph(directed = false) {
  const nodes = [];
  const edges = [];

  return {
    directed,
    nodes,
    edges,
    getNode(key) {
      return nodes.find(node => node.key === key) || null;
    },
    addNode(key) {
      if (key == null || this.getNode(key) != null) return null;

      const newNode = createNode(key);
      nodes.push(newNode);

      return newNode;
    },
    addEdge(node1Key, node2Key) {
      const [node1, node2] = [this.getNode(node1Key), this.getNode(node2Key)];

      if (!node1 || !node2) return null;

      node1.addNeighbor(node2);
      if (!this.directed) node2.addNeighbor(node1);

      const newEdge = `${node1Key}-${node2Key}`;

      edges.push(newEdge);

      return newEdge;
    },
    print() {
      return nodes.map(node => node.print()).join("\n");
    },
    traverseBreadth(startingKey, action) {
      const startingNode = this.getNode(startingKey);
      const visited = {};
      let currentNode;
      const queue = createQueue();
      if (startingNode) queue.enqueue(startingNode);
      else throw new Error(`Starting node not found`);

      while (!queue.isEmpty) {
        currentNode = queue.dequeue();
        currentNode.neighbors.forEach(node => {
          if (!visited[node.key]) queue.enqueue(node);
        });
        action(currentNode);
        visited[currentNode.key] = true;
      }
    },
    traverseDepth(startingKey, action) {
      const startingNode = this.getNode(startingKey);
      const visited = {};

      function explore(node) {
        if (visited[node.key]) return;

        action(node);
        visited[node.key] = true;
        node.neighbors.forEach(neighbor => explore(neighbor));
      }

      if (startingNode) explore(startingNode);
      else throw new Error(`Starting node not found`);
    }
  };
}
