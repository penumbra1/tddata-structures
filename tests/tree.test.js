import { createNode, createTree } from "../src/tree";

const mockNode = (key, children = []) => ({
  key,
  children,
  addChild: expect.any(Function)
});

describe("Node", () => {
  let n;

  beforeEach(() => {
    n = createNode(1);
  });

  test("createNode should create an empty node", () => {
    expect(n).toEqual(mockNode(1));
  });

  test("addChild should add a new child node", () => {
    const two = n.addChild(2);

    expect(n.children).toEqual([two]);
    expect(two).toEqual(mockNode(2));
  });
});

describe("Tree", () => {
  let t;

  beforeEach(() => {
    t = createTree(1);
  });

  test("createTree should throw if no key is specified", () => {
    const invalidCreate = () => createTree();

    expect(invalidCreate).toThrow();
  });

  test("createTree should create and return a tree with the specified root", () => {
    expect(t).toEqual({
      root: mockNode(1),
      print: expect.any(Function),
      traverse: expect.any(Function)
    });
  });

  test("print should return the correct string representation", () => {
    expect(t.print()).toBe("|--1\n");

    const two = t.root.addChild(2);
    const three = two.addChild(3);
    two.addChild(4);
    three.addChild(5);

    expect(t.print()).toMatchSnapshot();
  });
});
