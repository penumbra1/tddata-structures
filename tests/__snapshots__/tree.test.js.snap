// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree print should return the correct string representation 1`] = `
"|--1
|  |--2
|  |  |--3
|  |  |  |--5
|  |  |--4
"
`;
